### PR TITLE
Document Undocumented Functions in entry.h for Issue #371

### PR DESCRIPTION
**Pull request by JeanBaptisteDurand**
This pull request addresses Issue #371 by providing documentation comments for previously undocumented functions in the include/private/entry.h file. The following functions now have comments detailing their purpose and behavior:

locked_get_element_by_name
new_entry
strbuilder_append_app_name
strbuilder_append_hostname
strbuilder_append_msgid
strbuilder_append_message
strbuilder_append_procid
unchecked_destroy_entry
unchecked_entry_has_element
unlock_entry (Note: There's a relevant discussion left in the comments due to potential thread safety implications as discussed in Issue #136)

As my first ever contribution to an open-source project and pull request, I hope these changes will be beneficial and not waste your time. I have made an effort to ensure that each function is documented according to the project standards. There might be too much commits in this PR, reflecting each step of the documentation process.

Looking forward to feedback and suggestions.

Thank you.

### CODE REVIEW

**Comment by JeanBaptisteDurand** [comment]
Sorry for the inconvenience, I've rebase the commit to signed them

**Comment by goatshriek** [comment]
No worries, and thank you for the contribution! I will find the time to sit down and review this in the next couple of days.

File: include/private/entry.h [hunk 1]
```diff
@@ -29,7 +29,7 @@
 
 /**
  * Frees entry cache
- *
+ *stumpless_entry
```
Comment (goatshriek) [inline include/private/entry.h]: remove this stray addition

File: include/private/entry.h [hunk 2]
```diff
@@ -54,13 +54,7 @@ entry_free_all( void );
...
  * function with macros in "facility.h"
  *
- * **Thread Safety: MT-Safe**
- * This function is thread safe.
- *
- * **Async Signal Safety: AS-Safe **
- * This function must be safe to call from signal handlers
- *
- * **Async Cancel Safety: AC-Safe**
+ * **Thread Safety: MT-Safe[appropriate version]afe**
```
Comment (goatshriek) [inline include/private/entry.h]: This portion should be left as is (undo these changes).

File: include/private/entry.h [hunk 3]
```diff
@@ -148,14 +142,62 @@ struct stumpless_element *
...
                              size_t index );
 
+/**
+ * Retrieves an element by name from a Stumpless entry, ensuring thread safety.
+ *
+ * This function searches for an element within the given entry by its name. 
+ * It locks each element during comparison to ensure thread safety. If the 
+ * element with the specified name is found, it is returned.
+ *
+ * **Thread Safety: MT-Safe**
```
Comment (goatshriek) [inline include/private/entry.h]: `locked_` functions are not thread safe - review the naming conventions described in `docs/style.md`. In this case the function is MT-Unsafe as it operates on the element array of the entry without holding the lock, instead assuming the caller holds the appropriate lock.

File: include/private/entry.h [hunk 4]
```diff
@@ -148,14 +142,62 @@ struct stumpless_element *
...
+ * This function is thread-safe as it locks individual elements during the search process.
+ *
+ * **Async Signal Safety: AS-Unsafe**
+ * Unsafe to call from asynchronous signal handlers due to lock manipulation.
+ *
+ * **Async Cancel Safety: AC-Unsafe**
+ * Not safe in contexts of asynchronous cancellation, as it might involve lock manipulation.
+ *
+ * @since release 2.0.0
```
Comment (goatshriek) [inline include/private/entry.h]: Don't add `@since` tags in your additions - it's unnecessary work to go find the release that private functions were added in. If they're already present you can leave them, but don't worry about adding them to new documentation.

File: include/private/entry.h [hunk 5]
```diff
@@ -148,14 +142,62 @@ struct stumpless_element *
...
 
 /**
- * Creates a new entry with the given parameters.
+ * Creates a new Stumpless entry with specified parameters.
+ *
+ * This function allocates a new entry and initializes it with the provided facility,
+ * severity, application name, message ID, and message. It uses internal cache
+ * mechanisms for memory management and performs necessary validations and initializations.
+ *
+ * **Thread Safety: MT-Unsafe**
```
Comment (goatshriek) [inline include/private/entry.h]: This function is MT-Safe, though with race conditions on its string parameters. Memory allocation is assumed to be thread safe here, and the shared resources are okay to modify without synchronization as they are simply being initialized and haven't been shared with others yet.

File: include/private/entry.h [hunk 6]
```diff
@@ -165,21 +207,158 @@ new_entry( enum stumpless_facility facility,
...
+ *
+ * **Async Signal Safety: AS-Unsafe**
+ * Unsafe to call from asynchronous signal handlers due to memory manipulation
+ * and accessing shared data structures.
+ *
+ * **Async Cancel Safety: AC-Unsafe**
+ * Not safe in contexts of asynchronous cancellation, as it might lead to inconsistent
+ * states of shared data.
+ *
+ * @since release 1.0.0
```
Comment (goatshriek) [inline include/private/entry.h]: Remove this `@since` tag unless you've done the research to track down when this function appeared.

File: include/private/entry.h [hunk 7]
```diff
@@ -165,21 +207,158 @@ new_entry( enum stumpless_facility facility,
...
+ *
+ * **Async Signal Safety: AS-Unsafe**
+ * This function is not safe to call from asynchronous signal handlers as it
+ * involves system calls and memory manipulation.
+ *
+ * **Async Cancel Safety: AC-Unsafe**
+ * Not safe for use in contexts of asynchronous cancellation due to potential
+ * system call interruption and memory state inconsistency.
+ *
+ * @since release 1.0.0
```
Comment (goatshriek) [inline include/private/entry.h]: Similar `@since` tag removal here.

File: include/private/entry.h [hunk 8]
```diff
@@ -165,21 +207,158 @@ new_entry( enum stumpless_facility facility,
...
+ *
+ * **Async Signal Safety: AS-Unsafe**
+ * Unsafe to call from asynchronous signal handlers due to memory manipulation
+ * and accessing shared data structures.
+ *
+ * **Async Cancel Safety: AC-Unsafe**
+ * Not safe in contexts of asynchronous cancellation, as it might lead to inconsistent
+ * states of shared data.
+ *
+ * @since release 1.0.0
```
Comment (goatshriek) [inline include/private/entry.h]: Remove `@since` tag if not verified.

File: include/private/entry.h [hunk 9]
```diff
@@ -165,21 +207,158 @@ new_entry( enum stumpless_facility facility,
...
+ *
+ * **Async Signal Safety: AS-Unsafe**
+ * This function is not safe to call from asynchronous signal handlers as it
+ * involves memory manipulation and access to shared data structures.
+ *
+ * **Async Cancel Safety: AC-Unsafe**
+ * Not safe for use in contexts where threads may be asynchronously cancelled,
+ * as it could leave shared data in an inconsistent state.
+ *
+ * @since release 1.0.0
```
Comment (goatshriek) [inline include/private/entry.h]: Remove `@since` tag if not verified.

File: include/private/entry.h [hunk 10]
```diff
@@ -165,21 +207,158 @@ new_entry( enum stumpless_facility facility,
...
+ *
+ * **Async Signal Safety: AS-Unsafe**
+ * This function is not safe to call from asynchronous signal handlers as it
+ * relies on functions that may not be async-signal-safe.
+ *
+ * **Async Cancel Safety: AC-Unsafe**
+ * The function is not safe in the context of asynchronous cancellation due to
+ * potential resource cleanup issues.
+ *
+ * @since release v1.0.0
```
Comment (goatshriek) [inline include/private/entry.h:358]: Remove `@since` tag if not verified.

File: include/private/entry.h [hunk 11]
```diff
@@ -212,9 +391,55 @@ struct strbuilder *
...
+ *
+ * **Async Signal Safety: AS-Unsafe**
+ * Not safe to call from signal handlers as it involves operations like memory
+ * deallocation and mutex destruction.
+ *
+ * **Async Cancel Safety: AC-Unsafe**
+ * Not safe in the context of asynchronous cancellation due to potential resource
+ * cleanup issues.
+ *
+ * @since release v1.5.0
```
Comment (goatshriek) [inline include/private/entry.h]: Remove `@since` tag if not verified.

File: include/private/entry.h [hunk 12]
```diff
@@ -212,9 +391,55 @@ struct strbuilder *
...
+ *
+ * **Async Signal Safety: AS-Unsafe**
+ * This function is not safe to call from asynchronous signal handlers due to
+ * potential manipulation of shared data structures.
+ *
+ * **Async Cancel Safety: AC-Unsafe**
+ * This function is not safe for contexts where threads might be asynchronously
+ * cancelled, as it could leave shared data in an inconsistent state.
+ *
+ * @since release v1.6.0
```
Comment (goatshriek) [inline include/private/entry.h:438]: Remove `@since` tag if not verified.

File: include/private/entry.h [hunk 13]
```diff
@@ -296,7 +521,23 @@ unchecked_load_entry( struct stumpless_entry *entry,
...
+ * 
+ * This function is used internally to ensure that the mutex of an entry
+ * is properly unlocked after operations that required synchronization are
+ * completed. It uses the config_unlock_unchecked_entry_has_element
+ * **Async Cancel Safety: AC-Unsafe**
+ * This function is not safe to call from threads that may be asynchronously
+ * cancelled, as the cleanup of the lock may not be completed if a cancellation
+ * request is received during execution.
+ * 
+ * @since release v2.0.0
```
Comment (goatshriek) [inline include/private/entry.h:541]: Remove `@since` tag if not verified.

File: include/private/entry.h [hunk 14]
```diff
@@ -296,7 +521,23 @@ unchecked_load_entry( struct stumpless_entry *entry,
 void
 unchecked_unload_entry( const struct stumpless_entry *entry );
 
+/**
+ * Unlocks the mutex of a given entry.
+ * 
+ * This function is used internally to ensure that the mutex of an entry
+ * is properly unlocked after operations that required synchronization are
+ * completed. It uses the config_unlock_unchecked_entry_has_element
+ * **Async Cancel Safety: AC-Unsafe**
```
Comment (goatshriek) [inline include/private/entry.h:537]: The other two thread safety concepts are missing in this documentation block.

**Review by goatshriek — CHANGES_REQUESTED** [review]
First of all, thank you very much for putting these together! Tackling thread safety concepts can be tricky, and you did a great job of both summarizing functionality and describing the safety properties of these functions.

There are some changes below for you to make, after which this will be pulled in.

I have marked the `@since` tags as something you can simply delete if you don't know when the function was added. It is a lot of work to go back through commit histories to verify, and not worth the time to do this. If you've already done this work then leave them in of course, but otherwise you can simply leave them out.

**Comment by JeanBaptisteDurand** [comment]
Thank you for your time and feedback. I have made the requested changes. I really enjoy working on this project and hope to continue contributing. I hope to be useful and am learning a lot by reviewing the project's code.

File: include/private/entry.h [hunk 15]
```diff
@@ -122,40 +122,91 @@ locked_add_element( struct stumpless_entry *entry,
...
+ * Retrieves an element by name from a Stumpless entry, ensuring thread safety.
  *
- * **Thread Safety: MT-Unsafe**
- * This function is not thread safe as it doesn't use any synchronization 
- * or locking mechanisms while accessing shared resources.
+ * This function searches for an element within the given entry by its name.
+ * It locks each element during comparison to ensure thread safety. If the
+ * element with the specified name is found, it is returned.
+ *
+ * **Thread Safety: MT-Safe**
```
Comment (goatshriek) [inline include/private/entry.h]: This function is MT-Unsafe for the same reason as `locked_get_element_by_name`.

**Review by goatshriek — CHANGES_REQUESTED** [review]
Just two changes remain, correcting the safety of one function, and making sure that `lock_entry` has all three safety concepts documented.

**Comment by goatshriek** [comment]
Also, don't worry about the CI failures - they are a codecov issue that can be fixed by re-running and aren't something you caused.

**Comment by JeanBaptisteDurand** [comment]
I've made further revisions to the unlock_entry function, incorporating my understanding of the config_unlock_mutex macro, which is why there are two commits. Initially uncertain about its operation, I now understand that it aligns differently depending on the system, clarifying its role in thread synchronization. I have updated the documentation accordingly and committed these changes. I hope this revision are fine.

File: include/private/entry.h [hunk 16]
```diff
@@ -199,22 +395,80 @@ strbuilder_append_procid( struct strbuilder *builder );
...
+ * **Async Cancel Safety: AC-Unsafe**
+ * This function is not safe to call from threads that may be asynchronously
+ * cancelled, due to the use of lock that could be left in undefined state.
  *
- * @return The modified builder if no error is encountered. If an error is
+ * @param entry The entry to be locked.
+ *
+ */
+void
+lock_entry( const struct stumpless_entry *entry );
```
Comment (goatshriek) [inline include/private/entry.h]: looks like there are multiple declarations of `lock_entry` now? This should be resolved.

File: include/private/entry.h [hunk 17]
```diff
@@ -122,40 +122,101 @@ locked_add_element( struct stumpless_entry *entry,
...
 /**
- * Creates a new entry with the given parameters.
+ * Creates a new Stumpless entry with specified parameters.
+ *
+ * Allocates a new entry and initializes it with the provided facility,
+ * severity, application name, message ID, and message. It assumes that memory
+ * allocation (e.g., for the entry itself and string copies) is thread-safe as
+ * provided by the system's standard library.
+ */**
+ * Locks the mutex within the entry.
```
Comment (goatshriek) [inline include/private/entry.h]: It looks like the documentation for `lock_entry` and `create_entry` got interleaved somehow?

File: include/private/entry.h [hunk 18]
```diff
@@ -165,21 +226,156 @@ new_entry( enum stumpless_facility facility,
...
+ * This function is not safe to call from threads that may be asynchronously
+ * cancelled, due to the use of lock that could be left in undefined state.
+ *
+ * @param entry The entry to be locked.
+ *
+ */
+void
+lock_entry( const struct stumpless_entry *entry );
+ * Appends the process ID to the string builder.
+ *
```
Comment (goatshriek) [inline include/private/entry.h]: more interleaving here

**Review by goatshriek — CHANGES_REQUESTED** [review]
It looks like some comments and declarations got interleaved somehow, perhaps through a mistaken copy/paste? These will need to be cleaned up.

**Comment by JeanBaptisteDurand** [comment]
For the lock_entry function that appear everywhere... It seems like my commit from another PC and then trying to sign it on my main one did something weird. My bad on that. Just fixed it up, so it should be all good now. Let me know if anything still looks off!
Thanks for your patience

**Comment by goatshriek** [comment]
This is very close! Just one more minor adjustment to fix the [static analysis](https://github.com/goatshriek/stumpless/actions/runs/7519896826/job/20533934343?pr=399): `@since` tags need to follow the format "release vx.x.x". The current versions you've added are missing the `v` prefix to the version.

**Comment by JeanBaptisteDurand** [comment]
Should be good now

**Review by goatshriek — APPROVED** [review]

**Comment by goatshriek** [comment]
Thank you very much for putting this together, and sticking with it through the requested changes!
